src/battle_util.c

```diff
@@ -4335,6 +4295,28 @@ static u8 ForewarnChooseMove(u32 battler)
     Free(data);
 }
 
+void ChangeTypeBasedOnTerrain(u8 battlerId)
```

Comment:
I suggest making this return a `bool32`, and the failsafe option should return `FALSE`

Reply:
Uh, sure, I guess. It feels kinda weird to me, but 🤷 
```c
bool8 ChangeTypeBasedOnTerrain(u8 battlerId)
{
    u8 battlerType;
    u16 terrainFlags = VarGet(VAR_TERRAIN) & STATUS_FIELD_TERRAIN_ANY;

    if (terrainFlags && gFieldStatuses & STATUS_FIELD_TERRAIN_ANY)
        gFieldStatuses = terrainFlags | STATUS_FIELD_TERRAIN_PERMANENT; // terrain is permanent

    if (gFieldStatuses & STATUS_FIELD_ELECTRIC_TERRAIN)
        battlerType = TYPE_ELECTRIC;
    else if (gFieldStatuses & STATUS_FIELD_GRASSY_TERRAIN)
        battlerType = TYPE_GRASS;
    else if (gFieldStatuses & STATUS_FIELD_MISTY_TERRAIN)
        battlerType = TYPE_FAIRY;
    else if (gFieldStatuses & STATUS_FIELD_PSYCHIC_TERRAIN)
        battlerType = TYPE_PSYCHIC;
    else // failsafe
        return FALSE;
    SET_BATTLER_TYPE(battlerId, battlerType);
    PREPARE_TYPE_BUFFER(gBattleTextBuff1, battlerType);
}
```